### PR TITLE
For#27697 - Set unified search toolbar icon before the screen is visible

### LINKED ISSUE #27697: Unified search delay in showing icon when switching between display mode and edit mode
https://user-images.githubusercontent.com/6579541/199828530-48ff26b1-2a3e-461c-ba10-ab38d0e9192a.mp4

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21927)
┆Link To Issue: https://mozilla-hub.atlassian.net/browse/FNXV2-21927

### PR DESCRIPTION
**Pull request by DreVla**
Before Video:

https://user-images.githubusercontent.com/32488956/200375376-ab40c5f4-5124-48b5-98b6-a99640e96f35.mp4

After Video

https://user-images.githubusercontent.com/32488956/200375426-680b643d-7f8c-4dd5-8c23-4cbcd87e0ffe.mp4

As you can see in the video, the fix should work for the bookmarks and history toolbar as well.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [X] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #27697

### CODE REVIEW

**Comment by rocketsroger** [comment]
@DreVla thanks for the PR.  If you don't mind can we first wait till [this](https://github.com/mozilla-mobile/fenix/pull/27647) is landed before reviewing your change? I'll would like to test your change with that PR to make sure it still works.  Thanks,

**Comment by rocketsroger** [comment]
@DreVla The [change](https://github.com/mozilla-mobile/fenix/pull/27647) that was blocking your PR has landed.  Thanks

**Comment by MatthewTighe** [comment]
👋 I'm happy to take a look at this while Roger is on PTO, but I'm curious about the state of it. The [designs](https://www.figma.com/file/epu0Sj6mCPk7K282sNpHgI/Site?node-id=3851%3A77540&t=fKr0m7FT40kBlDHB-0#296011078) seem to show the selector as always visible but that does not appear to be the case in this patch. With this patch it seems the selector is only available while the toolbar is in edit mode.

Additionally, I'm only able to detect any visual stuttering on my Moto G5, which is a very slow device. I do see improvement with this patch, but I am curious if that will continue to be the case once the selector is always shown.

It also looks like the test failure may be a real one.

**Comment by DreVla** [comment]
@MatthewTighe I am currently still looking into this PR. You were right about the selector only showing in edit mode, but that issue seems to be fixed. As for the stuttering it is noticeable on a Nexus 5 and barely noticeable on a OnePlus 6. It is way more noticeable when unified search is turned off since the hint moves in the toolbar.

https://user-images.githubusercontent.com/32488956/207362630-fe9af6f4-70ab-43f8-a817-75209e528696.mp4

As for the test, It seems like now it is fixed, but I am encountering a new bug, where the engine is not changed if tapped when toolbar is not in edit mode:

https://user-images.githubusercontent.com/32488956/207363426-d4825948-c1fb-4424-bf97-18d3796b0a42.mp4

So this will probably take some more investigation..

**Review by MatthewTighe — APPROVED** [review]
Couple small things, but this otherwise LGTM!

File: app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorToolbarAction.kt [hunk 1]
```diff
@@ -36,14 +36,26 @@ import org.mozilla.fenix.search.SearchDialogFragmentStore
...
     private val store: SearchDialogFragmentStore,
+    private val defaultSearchEngine: SearchEngine?,
     private val menu: SearchSelectorMenu,
 ) : Toolbar.Action {
     private var updateIconJob: Job? = null
 
     override fun createView(parent: ViewGroup): View {
         val context = parent.context
 
+        // Only the application search engines (tabs etc) will have a valid icon at this time
```
[the remark this answers is not in the document]
Comment (DreVla) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorToolbarAction.kt]: I have changed this comment to a more clearer one. The engines I was referring to are tabs, bookmarks, history, which are of type APPLICATION, unlike the BUNDLED (google, amazon, bing, etc...) or CUSTOM ones.

[the remark this answers is not in the document]
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/search/toolbar/SearchSelectorToolbarAction.kt]: That seems much more clear to me, thank you 🙂

**Pull request merged**